### PR TITLE
Logger cleanup

### PR DESCRIPTION
**Pull request by pepone**
This small PR cleanup logger classes

* Replace GetPrefix method by Prefix property
* Rework locks, the console logger use an static global lock, and the file logger use a instance lock
* Added doc comments to public logger classes, sort fields, ...

### CODE REVIEW

File: csharp/src/Ice/Logger.cs [hunk 1]
```diff
@@ -8,161 +8,123 @@
...
+            _file = file;
+            _mutex = mutex;
+            _writer = writer;
         }
+    }
 
-        public abstract ILogger CloneWithPrefix(string prefix);
-
-        protected abstract void Write(string message);
+    public abstract class Logger : ILogger
```
Comment (bernardnormier) [inline csharp/src/Ice/Logger.cs:76]: I would keep this class at the top of the file, before its derived classes.

File: csharp/src/Ice/Logger.cs [hunk 2]
```diff
@@ -8,161 +8,123 @@
...
-
-        public override void Write(string message) => System.Console.Error.Write(message);
-
-        public override void WriteLine(string message) => System.Console.Error.WriteLine(message);
-
-        private const string Date = "d";
-        private const string Time = "HH:mm:ss:fff";
     }
 
     public sealed class TraceLogger : Logger
```
Comment (bernardnormier) [inline csharp/src/Ice/Logger.cs:123]: What's the purpose of a TraceLogger?

File: csharp/src/Ice/Logger.cs [hunk 3]
```diff
@@ -8,161 +8,123 @@
...
+            _file = file;
+            _mutex = new object();
+            _writer = new StreamWriter(new FileStream(file, FileMode.Append, FileAccess.Write, FileShare.Read));
         }
 
-        public virtual void Error(string message)
+        public override ILogger CloneWithPrefix(string prefix) => new FileLogger(prefix, _file, _mutex, _writer);
+
+        /// <summary>Close the underlying stream an dispose all the resources associated with it.</summary>
+        public void Destroy() => _writer.Close();
```
Comment (bernardnormier) [inline csharp/src/Ice/Logger.cs:55]: Make FileLogger IDisposable and remove Destroy?

File: csharp/src/Ice/Logger.cs [hunk 4]
```diff
@@ -8,161 +8,123 @@
...
         }
+    }
 
-        public virtual void Warning(string message)
+    /// <summary>Represents a logger that writes messages to a file.</summary>
+    public sealed class FileLogger : Logger
+    {
+        private readonly string _file;
+        private readonly object _mutex;
+        private readonly TextWriter _writer;
```
Comment (bernardnormier) [inline csharp/src/Ice/Logger.cs:39]: Should we use instead a synchronized TextWriter and eliminate the mutex?

File: csharp/src/Ice/Logger.cs [hunk 5]
```diff
@@ -211,8 +169,40 @@ private string Format(string category, string message)
...
+        private const string Time = "HH:mm:ss:fff";
+
+        public override void TraceEvent(
+            TraceEventCache cache,
+            string source,
+            TraceEventType type,
+            int id,
+            string message)
+        {
+            var s = new System.Text.StringBuilder(
```
Comment (bernardnormier) [inline csharp/src/Ice/Logger.cs:188]: Add `using System.Text` in this file?

File: csharp/src/Ice/Logger.cs [hunk 6]
```diff
@@ -8,161 +8,123 @@
...
         }
 
-        public virtual void Trace(string category, string message)
+        public override ILogger CloneWithPrefix(string prefix) => new ConsoleLogger(prefix);
+
+        protected override void Write(string message)
         {
-            string s = Format("--", category, message);
-            lock (GlobalMutex)
+            lock (_globalMutex)
```
Comment (bernardnormier) [inline csharp/src/Ice/Logger.cs:27]: Is this locking necessary? System.Console.Error.WriteLine must be thread-safe.

Comment (pepone) [reply csharp/src/Ice/Logger.cs:27]: Right seems we can get rid of it, according to https://docs.microsoft.com/en-us/dotnet/api/system.console?view=netcore-3.1#console-io-streams

File: csharp/src/Ice/Logger.cs (hunk 2 above)
Comment (pepone) [reply csharp/src/Ice/Logger.cs:123]: This logger publish log messages using the Trace api, .NET send the message to all the listeners configured, we have a console listener that writes to the console, but users can add other listeners programmatically or with .NET configuration properties

https://docs.microsoft.com/en-us/dotnet/framework/debug-trace-profile/trace-listeners

We have  a property that controls when the console listener is added

https://doc.zeroc.com/ice/3.7/property-reference/miscellaneous-ice-properties#id-.MiscellaneousIce.*Propertiesv3.7-Ice.ConsoleListener

Comment (pepone) [reply csharp/src/Ice/Logger.cs:123]: The property documentation is a bit confusing, ` If num is 0, logging is disabled.`, this doesn't disable logging, when set to 0 the `ConsoleListener` that writes the messages to the standard error is not added to the trace listeners, but logger is still enabled and will be send to any listeners added by the application

File: csharp/src/Ice/Logger.cs (hunk 4 above)
Comment (pepone) [reply csharp/src/Ice/Logger.cs:39]: Yes, seems like a good idea

**Pull request merged**